sklearn/datasets/_openml.py

```diff
@@ -77,20 +77,69 @@ def _open_openml_url(openml_path, data_home):
 
     Returns
     -------
-    result : stream
-        A stream to the OpenML resource
+    result : bytes
+        Byte content of resource
     """
     def is_gzip(_fsrc):
         return _fsrc.info().get('Content-Encoding', '') == 'gzip'
 
     req = Request(_OPENML_PREFIX + openml_path)
     req.add_header('Accept-encoding', 'gzip')
 
+    def _md5_validated_bytestream(fsrc, expected_md5=None, chunk_size=512):
+        """
+        Takes in a byte-stream, reads in chunks and returns bytes.
+        If expected_md5 is not none, keeps md5 checksum state while streaming
+        and validates post stream consumption.
+
+        Parameters
+        ----------
+        fsrc : io.BufferedIOBase
+            input stream to read bytes from
+
+        expected_md5 : str
+            expected md5-checksum value
+
+        chunk_size : int
+            size of chunks to read at a time from stream
+
+        Returns
+        -------
+        fsrc_bytes : bytes
+            equivalent to fsrc_bytes.read() but with md5 validation if
+            expected_md5 is provided
+
+        Raises
+        ------
+        ValueError :
+            if expected_md5 does not match actual md5-checksum of stream
+        """
+        if expected_md5 is None:
+            return fsrc.read()
+
+        fsrc_bytes = bytearray()
+        file_md5 = hashlib.md5()
+        while True:
+            data = fsrc.read(chunk_size)
+            if not data:
+                break
+            file_md5.update(data)
+            fsrc_bytes.extend(data)
```

Comment:
Is this where the extra memory comes in? Should we instead be writing straight to fdst in this function, rather than accumulating the zipped content in memory?

Reply:
So we are keeping a 512byte buffer into which we keep reading from the response / gzip file. I think there should only be a 512KB extra memory kept (apart from the whole buffer assuming all interim objects were garbage collected, worst case this could be twice the size of the stream)

Your point brings an even more important question, the md5checksum of a URL present in openml get_data_id api, if the url is a .gz would the md5sum also be of the .gz and not uncompressed?

If that is the case then I need to address that in the PR, perhaps I should validate on a stream as is (and not create a GzipFile.read)

Reply:
we should always be digesting the uncompressed data... But perhaps we already are.

> So we are keeping a 512byte buffer into which we keep reading from the response

Is fsrc_bytes really limited to 512B?

Reply:
> we should always be digesting the uncompressed data... But perhaps we already are.

Thanks. I guess then the openml get_data_id api gives digest of the uncompressed file always...

> Is fsrc_bytes really limited to 512B?

I was talking about data, the temp var being 512B, which is used to read fsrc. So right now we are creating a copy of the arff byte content from fsrc.read() -> fsrc_bytes after calculating bytes checksum.

@thomasjpfan Is the extra memory because even though we are reading the stream once we are [copying](https://stackoverflow.com/questions/35880065/does-converting-from-bytearray-to-bytes-incur-a-copy) it essentially to be able to find checksum yet return a stream?